### ISSUE TITLE
Silence in a very specific situation

### ISSUE THREAD

**Issue by renezuidhof** [superpoweredSDK/Low-Latency-Android-iOS-Linux-Windows-tvOS-macOS-Interactive-Audio-Platform#725]
For some reason all audio in my Android app got silenced when I was using some specific FXs. It took a bit to find what was actually causing it, but I was able to reproduce this by editing the `SuperpoweredPlayer` example project for Android.
Below you can find the edited file. Besides that the only thing I changed in the project was the ndkPath in gradle file.

It seems that when you have this situation: `Superpowered3BandEQ` > `SuperpoweredTimeStretching` (just pitch shifting, no stretching) > `SuperpoweredReverb` the audio gets silenced. But not directly, only once the first silence occurs.

When you run this file the audio will play including the FX with no problems. But once you pause, there is no audio anymore. Also after playing the file again. If you would remove any of the FX, things work fine. It's just in this specific situation that this happens for me.

This problem does not occur when I'm running it on an emulator. I only have it on actual Android devices (tested it on two phones).

Any idea why this might be happening?

I'm using Superpowered 2.5.1.

`PlayerExample.cpp`
```
#include <jni.h>
#include <string>
#include <android/log.h>
#include <OpenSource/SuperpoweredAndroidAudioIO.h>
#include <Superpowered.h>
#include <SuperpoweredAdvancedAudioPlayer.h>
#include <SuperpoweredSimple.h>
#include <SuperpoweredCPU.h>
#include <malloc.h>
#include <SLES/OpenSLES_AndroidConfiguration.h>
#include <SLES/OpenSLES.h>
#include <SuperpoweredTimeStretching.h>
#include <SuperpoweredReverb.h>
#include <SuperpoweredMixer.h>
#include <Superpowered3BandEQ.h>

#define log_print __android_log_print

static SuperpoweredAndroidAudioIO *audioIO;
static Superpowered::AdvancedAudioPlayer *player;
static Superpowered::TimeStretching *timeStretching;
static Superpowered::Reverb *reverb;
static Superpowered::ThreeBandEQ *eq;

// This is called periodically by the audio engine.
static bool audioProcessing (
        void * __unused clientdata, // custom pointer
        short int *audio,           // output buffer
        int numberOfFrames,         // number of frames to process
        int samplerate              // current sample rate in Hz
) {
    float mainAudioFloatArray[numberOfFrames * 2];

    for (int i = 0; i < numberOfFrames * 2; i++) {
        mainAudioFloatArray[i] = 0.0f;
    }

    player->processStereo(mainAudioFloatArray, false, (unsigned int)numberOfFrames);

    eq->process(mainAudioFloatArray, mainAudioFloatArray, numberOfFrames);

    timeStretching->addInput(mainAudioFloatArray, (int)numberOfFrames);
    timeStretching->getOutput(mainAudioFloatArray, (int)numberOfFrames);

    reverb->process(mainAudioFloatArray, mainAudioFloatArray, (unsigned int)numberOfFrames);

    Superpowered::FloatToShortInt(mainAudioFloatArray, audio, (unsigned int) numberOfFrames);

    return true;
}

// StartAudio - Start audio engine and initialize player.
extern "C" JNIEXPORT void
Java_com_superpowered_playerexample_MainActivity_NativeInit(JNIEnv *env, jobject __unused obj, jint samplerate, jint buffersize, jstring tempPath) {
    Superpowered::Initialize("ExampleLicenseKey-WillExpire-OnNextUpdate");

    // setting the temp folder for progressive downloads or HLS playback
    // not needed for local file playback
    const char *str = env->GetStringUTFChars(tempPath, 0);
    Superpowered::AdvancedAudioPlayer::setTempFolder(str);
    env->ReleaseStringUTFChars(tempPath, str);

    // creating the player
    player = new Superpowered::AdvancedAudioPlayer((unsigned int)samplerate, 0);

    eq = new Superpowered::ThreeBandEQ((unsigned int) samplerate);
    eq->enabled = true;

    reverb = new Superpowered::Reverb((unsigned int)samplerate);
    reverb->enabled = true;

    timeStretching = new Superpowered::TimeStretching((unsigned int)samplerate, 1.0F);
    timeStretching->pitchShiftCents = 834; // < Random number

    audioIO = new SuperpoweredAndroidAudioIO (
            samplerate,                     // device native sampling rate
            buffersize,                     // device native buffer size
            false,                          // enableInput
            true,                           // enableOutput
            audioProcessing,                // process callback function
            NULL,                           // clientData
            -1,                             // inputStreamType (-1 = default)
            SL_ANDROID_STREAM_MEDIA         // outputStreamType (-1 = default)
    );
}

// OpenFile - Open file in player, specifying offset and length.
extern "C" JNIEXPORT void
Java_com_superpowered_playerexample_MainActivity_OpenFileFromAPK (
        JNIEnv *env,
        jobject __unused obj,
        jstring path,       // path to APK file
        jint offset,        // offset of audio file
        jint length         // length of audio file
) {
    const char *str = env->GetStringUTFChars(path, 0);
    player->open(str, offset, length);
    env->ReleaseStringUTFChars(path, str);

    // open file from any path: player->open("file system path to file");
    // open file from network (progressive download): player->open("http://example.com/music.mp3");
    // open HLS stream: player->openHLS("http://example.com/stream");
}

// onUserInterfaceUpdate - Called periodically. Check and react to player events. This can be done in any thread.
extern "C" JNIEXPORT jboolean
Java_com_superpowered_playerexample_MainActivity_onUserInterfaceUpdate(JNIEnv * __unused env, jobject __unused obj) {
    switch (player->getLatestEvent()) {
        case Superpowered::AdvancedAudioPlayer::PlayerEvent_None:
        case Superpowered::AdvancedAudioPlayer::PlayerEvent_Opening: break; // do nothing
        case Superpowered::AdvancedAudioPlayer::PlayerEvent_Opened: player->play(); break;
        case Superpowered::AdvancedAudioPlayer::PlayerEvent_OpenFailed:
        {
            int openError = player->getOpenErrorCode();
            log_print(ANDROID_LOG_ERROR, "PlayerExample", "Open error %i: %s", openError, Superpowered::AdvancedAudioPlayer::statusCodeToString(openError));
        }
            break;
        case Superpowered::AdvancedAudioPlayer::PlayerEvent_ConnectionLost:
            log_print(ANDROID_LOG_ERROR, "PlayerExample", "Network download failed."); break;
        case Superpowered::AdvancedAudioPlayer::PlayerEvent_ProgressiveDownloadFinished:
            log_print(ANDROID_LOG_ERROR, "PlayerExample", "Download finished. Path: %s", player->getFullyDownloadedFilePath()); break;
    }

    if (player->eofRecently()) player->setPosition(0, false, false);
    return (jboolean)player->isPlaying();
}

// TogglePlayback - Toggle Play/Pause state of the player.
extern "C" JNIEXPORT void
Java_com_superpowered_playerexample_MainActivity_TogglePlayback(JNIEnv * __unused env, jobject __unused obj) {
    player->togglePlayback();
    Superpowered::CPU::setSustainedPerformanceMode(player->isPlaying()); // prevent dropouts
}

// onBackground - Put audio processing to sleep if no audio is playing.
extern "C" JNIEXPORT void
Java_com_superpowered_playerexample_MainActivity_onBackground(JNIEnv * __unused env, jobject __unused obj) {
    audioIO->onBackground();
}

// onForeground - Resume audio processing.
extern "C" JNIEXPORT void
Java_com_superpowered_playerexample_MainActivity_onForeground(JNIEnv * __unused env, jobject __unused obj) {
    audioIO->onForeground();
}

// Cleanup - Free resources.
extern "C" JNIEXPORT void
Java_com_superpowered_playerexample_MainActivity_Cleanup(JNIEnv * __unused env, jobject __unused obj) {
    delete audioIO;
    delete player;
    delete reverb;
    delete timeStretching;
    delete eq;
}
```

**Comment by renezuidhof** [comment]
It seems to me that the timestretcher (used as a pitch-shifter) doesn't handle data with just `0.0f`'s well. Therefor passing bad data to reverb, which just stops working properly causing the issue. But I might be wrong there, just sharing my thoughts.

**Comment by gaborszanto** [comment]
The time stretcher getOutput may return with false, meaning it returned with no output. You forgot to handle that situation.

In your example the player has the time stretcher integrated, so you don't need a time stretcher instance at all.

**Comment by renezuidhof** [comment]
The example I posted is a simplified version. In my app I need the 'external' timestretcher to do pitchshifting.

I know getOutput may return false, but it doesn't. So this cannot be the reason for the audio going silence.

To test I added this:

```
if (!hasOutput) {
    log_print(ANDROID_LOG_ERROR, "PlayerExample", "timeStretcher has no output!");

    return false;
}
```

after

```
timeStretching->getOutput(mainAudioFloatArray, (int)numberOfFrames);
```

**Comment by renezuidhof** [comment]
As a (nasty) workaround I've changed the initialisation of the `mainAudioFloatArray` to this. In this case this bug does not occur.

```
mainAudioFloatArray[0] = 0.0001f;

for (int i = 1; i < numberOfFrames * 2; i++) {
    mainAudioFloatArray[i] = 0.0f;
}
```
